mobile_verifier/src/reward_speed_share.rs

```diff
@@ -0,0 +1,327 @@
+        }
+        self.is_valid = self.check_validity()
+    }
+
+    fn check_validity(&self) -> bool {
+        self.speed_shares.len() >= MIN_REQUIRED_SAMPLES
+            && self.average.download_speed_avg_bps >= MIN_DOWNLOAD
+            && self.average.upload_speed_avg_bps >= MIN_UPLOAD
+            && self.average.latency_avg_ms <= MAX_LATENCY
+    }
+}
+
+#[derive(Debug, Serialize, Clone)]
+pub struct SpeedShare {
+    pub pub_key: PublicKey,
+    pub timestamp: u64,
+    pub upload_speed: u64,
+    pub download_speed: u64,
+    pub latency: u32,
+    pub validity: ShareValidity,
+}
+
+impl From<SpeedShare> for SpeedShareProto {
+    fn from(ss: SpeedShare) -> Self {
+        Self {
+            pub_key: ss.pub_key.to_vec(),
+            timestamp: ss.timestamp,
+            upload_speed_bps: ss.upload_speed,
+            download_speed_bps: ss.download_speed,
+            latency_ms: ss.latency,
+            validity: ss.validity as i32,
+        }
+    }
+}
+
+impl SpeedShare {
+    pub fn new(
+        pub_key: PublicKey,
+        timestamp: u64,
+        upload_speed: u64,
+        download_speed: u64,
+        latency: u32,
+        validity: ShareValidity,
+    ) -> Self {
+        Self {
+            pub_key,
+            timestamp,
+            upload_speed,
+            download_speed,
+            latency,
+            validity,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use chrono::DateTime;
+    use std::str::FromStr;
+
+    fn parse_dt(dt: &str) -> u64 {
+        DateTime::parse_from_str(dt, "%Y-%m-%d %H:%M:%S %z")
+            .expect("unable_to_parse")
+            .timestamp() as u64
+    }
+
+    fn bytes_per_s(mbps: u64) -> u64 {
+        mbps * 125000
+    }
+
+    fn known_shares() -> Vec<SpeedShare> {
+        // This data is taken from the spreadsheet
+        // Timestamp	DL	UL	Latency	DL RA	UL RA	Latency RA	Pass?
+        // 2022-08-01 0:00:00	0	0	0	0.00	0.00	0.00	FALSE*
+        // 2022-08-01 6:00:00	150	20	70	75.00	10.00	35.00	FALSE
+        // 2022-08-01 12:00:00	118	10	50	89.33	10.00	40.00	FALSE
+        // 2022-08-01 18:00:00	112	30	40	95.00	15.00	40.00	FALSE
+        // 2022-08-02 0:00:00	90	15	10	94.00	15.00	34.00	FALSE
+        // 2022-08-02 6:00:00	130	20	10	100.00	15.83	30.00	TRUE
+        // 2022-08-02 12:00:00	100	10	30	116.67	17.50	35.00	TRUE
+        // 2022-08-02 18:00:00	70	30	40	103.33	19.17	30.00	TRUE
+
+        let gw_public_key =
+            PublicKey::from_str("1126cBTucnhedhxnWp6puBWBk6Xdbpi7nkqeaX4s4xoDy2ja7bcd")
+                .expect("unable to get pubkey");
+        let mut shares = vec![];
+
+        let s0 = SpeedShare::new(
+            gw_public_key.clone(),
+            parse_dt("2022-08-01 0:00:00 +0000"),
+            0,
+            0,
+            0,
+            Validity::Valid,
+        );
+        let s1 = SpeedShare::new(
+            gw_public_key.clone(),
+            parse_dt("2022-08-01 6:00:00 +0000"),
+            bytes_per_s(25),
```

Comment:
```suggestion
            bytes_per_s(20),
```

to match the example in the comment above